### PR TITLE
SOLR-16959: Make CoresLocator class configurable

### PR DESCRIPTION
**Pull request by pvcnt**
https://issues.apache.org/jira/browse/SOLR-16959

# Description

I would like to be able to hook into the process of discovering and persisting core descriptors. This allows to work with core descriptors not stored locally, or replicated elsewhere.

# Solution

For this purpose, I propose to make the CoresLocator instance configurable, pretty much like other services (e.g., ConfigSetService, some handlers) are already configurable. Attached PR creates a new "coresLocator" parameter in solr.xml in order to achieve this. Because it is created outside of the CoreContainer (it is needed to load a CoreContainer), constructor of a custom CoresLocator receives a NodeConfig object (and not a CoreContainer) that it can use to lookup other parameters, if required.

# Tests

I added new unit tests to validate the correct instantiation of a CoresLocator.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Review by murblanc — COMMENTED** [review]
Thanks for doing this Vincent. Looks good to me.

Full disclaimer: I work with Vincent in the same company and that change, in addition to aligning `CoresLocator` with configuration flexibility available for other classes will also eventually simplify the integration of new types of cores that have no local presence (index Blob/S3 storage Salesforce plans to eventually contribute, see [email](https://lists.apache.org/thread/3br63xjmv20dp61j242bj747g2msmdov) .

File: solr/core/src/java/org/apache/solr/core/NodeConfig.java [hunk 1]
```diff
@@ -154,6 +155,7 @@ private NodeConfig(
       Set<Path> allowPaths,
       List<String> allowUrls,
       String configSetServiceClass,
+      String coresLocatorClass,
```
[the remark this answers is not in the document]
Comment (pvcnt) [reply solr/core/src/java/org/apache/solr/core/NodeConfig.java:158]: Agreed, fixed.

File: solr/core/src/java/org/apache/solr/core/CoresLocator.java [hunk 2]
```diff
@@ -72,4 +73,26 @@ public interface CoresLocator {
...
+        Class<? extends CoresLocator> clazz =
+            nodeConfig.getSolrResourceLoader().findClass(coresLocatorClass, CoresLocator.class);
+        Constructor<? extends CoresLocator> constructor = clazz.getConstructor(NodeConfig.class);
+        return constructor.newInstance(nodeConfig);
+      } catch (Exception e) {
+        throw new RuntimeException(
+            "create CoresLocator instance failed, coresLocatorClass=" + coresLocatorClass, e);
+      }
+    } else {
+      return new CorePropertiesLocator(nodeConfig.getCoreRootDirectory());
```
[the remark this answers is not in the document]
Comment (pvcnt) [reply solr/core/src/java/org/apache/solr/core/CoresLocator.java:95]: Fixed.

File: solr/solr-ref-guide/modules/configuration-guide/pages/configuring-solr-xml.adoc [hunk 3]
```diff
@@ -99,6 +99,18 @@ For example, `<str name="configSetService">com.myorg.CustomConfigSetService</str
...
+`coresLocator`::
++
+[%autowidth,frame=none]
+|===
+|Optional |Default: `org.apache.solr.core.CoresPropertiesLocator`
+|===
++
+This attribute does not need to be set.
++
+If used, this attribute should be set to the FQN (Fully qualified name) of a class that inherits from `CoresLocator`, and you must provide a constructor with one parameter which the type is `org.apache.solr.core.NodeConfig`.
```
[the remark this answers is not in the document]
Comment (pvcnt) [reply solr/solr-ref-guide/modules/configuration-guide/pages/configuring-solr-xml.adoc:111]: Fixed. And fixed the same typo in another section (I copy/pasted from...).

**Review by dsmiley — APPROVED** [review]

**Comment by dsmiley** [comment]
Can you please add a CHANGES.txt note in the 9.4 section, probably in the "Improvements" section?  A brief sentence would do like "Make the internal CoresLocator implementation configurable in solr.xml"

**Comment by pvcnt** [comment]
@dsmiley That's done.

**Pull request merged**